nova_ecs/utils.ts

```diff
@@ -47,30 +50,192 @@ export function topologicalSortList(list: Sortable[]): Sortable[] {
  */
 export function topologicalSort<T>(graph: Map<T, Set<T>>): T[] {
-    const usedNodes = new Set<T>();
-    const sorted: T[] = [];
-
-    while (sorted.length < graph.size) {
-        const lastLength = sorted.length;
-        for (const [node, incomingEdges] of graph) {
-            // Skip nodes we've already added
-            if (usedNodes.has(node)) {
-                continue;
+    if (graph.size == 0){
+        return [];
+    }
+    const domain = new Set<T>(graph.keys());
+    const nonleafs = new Set<T>();
+    const placed = new Set<T>();
+    const remaining = new Set<T>();
+    const result:T[] = [];
+    for (let [k,v] of graph.entries()){
+        if (setDifference(setIntersection(v,domain),placed).size == 0){
+            result.push(k);
+            placed.add(k);
+        }else{
+            remaining.add(k);
+        }
+        for (let n of v){
+            if (domain.has(n)){
+                nonleafs.add(n)
+            }
+        }
+    }
+    if (result.length == 0){
+        const cyclicPart = trimToCyclic(graph);
+        throw new GraphCycleError("Graph has no roots (contains cycles)\n"+graphListString(cyclicPart));
+    }
+    if (remaining.size == 0){
+        return result;
+    }
+    //find start set for depth-first-search 
+    const leafs = new Set<T>();
+    for (let n of remaining){
+        if (!nonleafs.has(n)){
+            leafs.add(n);
+        }
+    }
+    if (leafs.size == 0){
+        const cyclicPart = trimToCyclic(graph);
+        throw new GraphCycleError("Graph has no antiroots (contains cycles)\n"+graphListString(cyclicPart));
+    }
+    function dfs(c:Set<T>,ancestors:Set<T>){
+        for (let n of c) {
+            if (domain.has(n)){
+                if (ancestors.has(n)){
+                    const cyclicPart = trimToCyclic(graph);
+                    throw new GraphCycleError("Graph contains cycle\n"+graphListString(cyclicPart));
+                }
+                ancestors.add(n);
+                const next = graph.get(n);
+                if (next) {
+                    dfs(setDifference(next,placed),ancestors);
+                }
+                ancestors.delete(n);
+                result.push(n);
+                placed.add(n);
+            }
+        }
+    }
+    dfs(leafs,new Set<T>());
+    return result;
+}
+
+function graphListString<T>(graph:Map<T, Set<T>>):string{
+    /* produces descriptive text of the style:
+        ┌─c1
+        │ b1
+        │ a1
+       ┌└>root
+       │
+       └──a2
+    */
+    const listing:T[] = [];
+    const indices = new Map<T,number>();
+    const seen = new Set<T>();
+    for (let k of graph.keys()){
+        if (seen.has(k)){
+            continue;
+        }
+        do {
+            seen.add(k);
+            indices.set(k,listing.length);
+            listing.push(k);
+            k = setDifference(graph.get(k)??new Set<T>(),seen).keys().next().value;
+        } while (k)
```

Comment:
Graph keys may be falsy. 
```suggestion
        } while (k != null)
```